File: exercises/small_problems/medium_1.rb
#1 Rotation Part 1
def rotate_array(arr)
  arr[1..-1] + [array[0]]
end

# Rotation Part 2
def rotate_rightmost_digits(num, n)
  array = num.to_s.chars
  new = array[0, array.size - n] + rotate_array(array[array.size - n, n])
  new.join.to_i
end

#3 Rotation Part 3
def max_rotation(num)
  start = num.to_s.size
  loop do
    num = rotate_rightmost_digits(num, start)
    start -= 1
    break if start < 2
  end
  num
end

#4 1000 Lights
def thousand_lights(num)
  lights = Hash.new()
  1.upto(num) { |i| lights[i] = false }
  1.upto(num) do |i|
    toggle_every_nth_light(lights, i)
  end
  lights.select { |k,v| v }.keys
end

def toggle_every_nth_light(lights, n)
  lights.each do |num, bool|
    lights[num] = !bool if num % n == 0
  end
  lights
end

#5 Diamonds!
def diamond(num)
  space = num / 2 + 1

  1.upto(num) do |i|
    if i.odd?
      space -= 1
      puts " " * space + "*" * i
    end
  end

  (num-2).downto(1) do |i|
    if i.odd?
      space += 1
      puts " " * space + "*" * i
    end
  end
end

#6 Stack Machine Interpretation
def minilang(str)
  stack = []
  register = 0

  commands = str.split(" ")
  commands.each do |cmd|
    case cmd
      when "PUSH"
        stack.push(register)
      when "ADD"
        register += stack.pop
      when "SUB"
        register -= stack.pop
      when "MULT"
        register *= stack.pop
      when "DIV"
        register /= stack.pop
      when "MOD"
        register = register % stack.pop
      when "POP"
        register = stack.pop
      when "PRINT"
        puts register
      else
        register = cmd.to_i
    end
  end
end

#7 Word to Digit
NUMBERS = {"zero"=> "0", "one"=> "1", "two"=> "2", "three"=> "3", "four"=> "4", "five"=> "5", "six"=> "6", "seven"=> "7", "eight"=> "8", "nine"=> "9"}

def word_to_digit(str)
  NUMBERS.keys.each do |word|
    str.gsub!(/\b#{word}\b/, NUMBERS[word])
  end
  str
end

#8 Fibonacci NUMBERS (Recursion)
def fibonacci(n)
  return 1 if n < 3
  fibonacci(n-1) + fibonacci(n-2)
end

#9 Fibonacci Numbers (Procedural)
def fibonacci(nth)
  first, last = [1, 1]
  3.upto(nth) do
    first, last = [last, first + last]
  end
  last
end

#10 Fibonacci Numbers (Last Digit)
def fibonacci_last(nth)
  fibonacci(nth).to_s[-1].to_i
end

File: js101_foundations/exercises/medium_1.js
/* eslint-disable */
// Rotation (Part 1)
const rotateArray = array => {
  let result = array;
  if (typeof result !== "object") {
    return undefined;
  } else if (result.length === 0) {
    return [];
  } else {
    return result.slice(1).concat(result[0]);
  }
}

// Rotation (Part 2)
const rotateString = str => {
  return str.slice(1) + str[0];
}

const rotateRightmostDigits = (num, count) => {
  let string = String(num);
  let left = string.slice(0, string.length - count);
  let right = string.slice(string.length - count);
  right = rotateString(right);
  return Number(left + right);
}

// Rotation (Part 3)
const maxRotation = num => {
  let number = String(num);
  
  for(let i = 0; i < number.length; i++) {
    num = rotateRightmostDigits(num, number.length - i);
  }
  return num;
}

// Stack Machine Interpretation
const minilang = str => {
  let stack = [];
  let register = 0;

  str.split(" ").forEach(el => {
    switch (el) {
      case "PUSH":
        stack.push(register);
        break;
      case "ADD":
        register += stack.pop();
        break;
      case "SUB":
        register -= stack.pop();
        break;
      case "MULT":
        register *= stack.pop();
        break;
      case "DIV":
        register = Math.floor(register / stack.pop());
        break;
      case "MOD":
        register = register % stack.pop();
        break;
      case "POP":
        register = stack.pop();
        break;
      case "PRINT":
        console.log(register);
        break;
      default: 
        register = Number(el);
    }
  })
}

// Word to Digit
const words = {
    zero:  0,
    one:   1,
    two:   2,
    three: 3,
    four:  4,
    five:  5,
    six:   6,
    seven: 7,
    eight: 8,
    nine:  9
  };

const wordToDigit = str => {
  Object.keys(words).forEach(word => {
    let regex = new RegExp(word, 'g');
    str = str.replace(regex, words[word]);
  })
  return str;
}

// Fibonacci Numbers (Recursion)
const fibonacci = num => {
  if (num < 3) {
    return 1;
  }

  return fibonacci(num - 1) + fibonacci(num - 2);
}

// Fibonacci Numbers (Procedural)
const fibonacci = num => {
  let previousTwo = [1, 1];

  for (let counter = 3; counter <= num; counter++) {
    previousTwo = [previousTwo[1], previousTwo[0] + previousTwo[1]];
  }
  return previousTwo[1];
}

// Fibonacci Numbers (Memoization)
let memo = {};
const fibonacci = num => {
  if (num < 3) {
    return 1;
  } else if (memo[num]) {
    return memo[num];
  } else {
    memo[num] = fibonacci(num - 1) + fibonacci(num - 2);
    return memo[num];
  }
}

